### DISCUSSION TITLE
ESP32-S2 pin won't hold during deepsleep

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#14295] [Using MicroPython] [answered]
I'm in the Pins & GPIO section of https://docs.micropython.org/en/latest/esp32/quickref.html trying to figure out how to hold an esp32-S2 pin during deepsleep. The code runs, I can see pin 34 go low on the multimeter during the 5s sleep, 
```
>>> modpwr=Pin(34, Pin.OUT, value=0, hold=True); modpwr(); sleep(5); modpwr(); deepsleep(30)
0
0
```
but the pin seems to float as soon as the deepsleep starts and the lte modem power pullup drags it up to 5v. Can some knowledgeable soul please explain what I need to do to get pin34 to hold low during deepsleep?

**Comment by davefes** [comment]
Found this somewhere on the internet:
[GPIO during deepsleep.docx](https://github.com/micropython/micropython/files/14965019/GPIO.during.deepsleep.docx)
Maybe pin 34 is a bad choice?

**Reply by kjm1102** [reply]
Tnx Dave but the way I read that doc is that since 34 is an RTC pin it should be holdable "Output-configured RTC pins will also retain their output direction and level in deep-sleep if pad hold is enabled with the hold=True argument"

**Reply by davefes** [reply]
OK.  Maybe, I focused too much on the pull-up or pull-down configuration.

I was recently surprised that it was even possible to do this ... I thought pin-hold only worked for lightsleep.

Good luck!

**Reply by davefes** [reply]
What happens if you put 10K to ground on pin34?

**Reply by kjm1102** [reply]
If I load it with a led to 0v it drops to 1v7, confirming my hunch that during/post deepsleep it has defaulted to an input instead of holding.

**Reply by davefes** [reply]
Could it be an output that has gone high?  How does a ESP32 Generic behave?

**Reply by kjm1102** [reply]
Nah, the fact that pin 34 floats up to the modems 5v means the pin has gone high impedance. I tried your docs esp32.gpio_deep_sleep_hold(True) for a non rtc pin (17) & it doesn't work either so looks like hold during deepsleep is broken on an S2. Works as advertised for generic esp32.

**Reply by kjm1102** [reply]
According to the datasheet pin34 on the S2 is in the VDD3P3_CPU / VDD_SPI power domain so your esp32.gpio_deep_sleep_hold(True) for a non rtc pin should have worked?

**Reply by davefes** [reply]
Sorry, I have never tried to hold GPIO in deepsleep.  
https://forum.micropython.org/viewtopic.php?t=7470
> Only the pins in the RTC power domain can hold during deep sleep because the cpu power domain, which includes the "digital I/O" pins, is turned off. Those latter pins continue the hold when power comes back on.

https://github.com/micropython/micropython/issues/8283

Finally found where I got it from ... Deepsleep in the QuickRef !!

**Reply by kjm1102** [reply]
The 'Only the pins in the RTC power domain can hold during deep sleep' statement is wrong. The 'configuration of non-RTC pins - including output level - can be retained by enabling pad hold on the pin and enabling GPIO pad hold during deep-sleep by calling esp32.gpio_deep_sleep_hold(True):' from your original doc is true. I did
```
Pin(34, Pin.OUT, value=0, hold=1); esp32.gpio_deep_sleep_hold(1)
```
and the pin held low through & after deepsleep. So got there finally but I can't find that doc you sent me in the Deepsleep quick ref. Do you have a link I could follow?

**Reply by davefes** [reply, marked as answer]
https://docs.micropython.org/en/latest/esp32/quickref.html

**Reply by kjm1102** [reply]
much obliged
